[vtk] Include <iterator> in DataModel header

<iterator> is needed for `std::inserter`. This didn't show up until Drake
started using -std=c++23.

--- Common/DataModel/vtkDataSetAttributesFieldList.cxx
+++ Common/DataModel/vtkDataSetAttributesFieldList.cxx
@@ -13,6 +13,7 @@
 #include <algorithm>
 #include <array>
 #include <functional>
+#include <iterator>
 #include <map>
 #include <set>
 #include <string>
